examples/raytracing/example_raytracing.cpp

```diff
@@ -0,0 +1,291 @@
+    auto const &c = sphere.centroid();
+    auto const r = sphere.radius();
+    return {{c[0] - r, c[1] - r, c[2] - r}, {c[0] + r, c[1] + r, c[2] + r}};
+  }
+};
+
+template <typename MemorySpace>
+struct Rays
+{
+  Kokkos::View<ArborX::Ray *, MemorySpace> _rays;
+};
+
+template <typename MemorySpace>
+struct ArborX::AccessTraits<Rays<MemorySpace>, ArborX::PredicatesTag>
+{
+  using memory_space = MemorySpace;
+
+  KOKKOS_FUNCTION static std::size_t size(const Rays<MemorySpace> &rays)
+  {
+    return rays._rays.extent(0);
+  }
+  KOKKOS_FUNCTION static auto get(Rays<MemorySpace> const &rays, std::size_t i)
+  {
+    return attach(intersects(rays._rays(i)), (int)i);
+  }
+};
+
+template <typename MemorySpace>
+struct AccumRaySphereInterDist
+{
+  using Vector = ArborX::Ray::Vector;
+
+  Kokkos::View<ArborX::Sphere *, MemorySpace> _spheres;
+  Kokkos::View<float *, MemorySpace> _accumulator;
+
+  KOKKOS_FUNCTION float dotProduct(Vector const &v1, Vector const &v2) const
+  {
+    return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
+  }
+
+  // Solves a*x^2 + b*x + c = 0.
+  // If a solution exists, return true and stores roots at x1, x2.
+  // If a solution does not exist, returns false.
+  KOKKOS_FUNCTION bool solveQuadratic(float const a, float const b,
+                                      float const c, float &x1, float &x2) const
+  {
+    assert(a != 0);
+
+    auto const discriminant = b * b - 4 * a * c;
+    if (discriminant < 0)
+      return false;
+    if (discriminant == 0)
+    {
+      x1 = x2 = -b / (2 * a);
+      return true;
+    }
+
+    // Instead of doing a simple
+    //    (-b +- std::sqrt(discriminant)) / (2*a)
+    // we use a more stable algorithm with less loss of precision (see, for
+    // example, https://www.scratchapixel.com/lessons/3d-basic-rendering/
+    // minimal-ray-tracer-rendering-simple-shapes/ray-sphere-intersection).
+    auto const sgn = [](float x) { return (int)(x > 0) - (int)(x < 0); };
+    auto const q = 0.5 * (-b - sgn(b) * std::sqrt(discriminant));
+    x1 = q / a;
+    x2 = c / q;
+
+    return true;
+  }
+
+  // Ray-Sphere intersection algorithm.
+  //
+  // The sphere can be expressed as the solution to
+  //     |p - c|^2 - r^2 = 0,           (1)
+  // where c is the center of the sphere, and r is the radius. On the other
+  // hand, any point on a bidirectional ray satisfies
+  //     p = o + t*d,                   (2)
+  // where o is the origin, and d is the direction vector.
+  // Substituting (2) into (1),
+  //     |(o + t*d) - c|^2 - r^2 = 0,   (3)
+  // results in a quadratic equation for unknown t
+  //     a2 * t^2 + a1 * t + a0 = 0
+  // with
+  //     a2 = |d|^2, a1 = 2*(d, o - c), and a0 = |o - c|^2 - r^2.
+  // Then, we only need to intersect the solution interval [tmin, tmax] with
+  // [0, +inf) for the unidirectional ray.
+  KOKKOS_FUNCTION float overlapDistance(ArborX::Ray const &ray,
+                                        ArborX::Sphere const &sphere) const
```

Comment:
```suggestion
  static KOKKOS_FUNCTION float overlapDistance(ArborX::Ray const &ray,
                                        ArborX::Sphere const &sphere)
```
?

Reply:
Don't think it matters. This function will only be called internally from a non-static one.

Reply:
Otherwise, it wouldn't even compile. :wink: